packages/playground/data-liberation/src/import/WP_Topological_Sorter.php

```diff
@@ -0,0 +1,461 @@
+	 */
+	protected $total_posts = 0;
+
+	/**
+	 * The current item being processed.
+	 */
+	protected $current_item = 0;
+
+	public function __construct( $options = array() ) {
+		if ( array_key_exists( 'session_id', $options ) ) {
+			$this->current_session = $options['session_id'];
+		}
+	}
+
+	/**
+	 * Get the name of the table.
+	 *
+	 * @return string The name of the table.
+	 */
+	public static function get_table_name() {
+		global $wpdb;
+
+		// Default is wp_{TABLE_NAME}
+		return $wpdb->prefix . self::TABLE_NAME;
+	}
+
+	/**
+	 * Run by register_activation_hook.
+	 */
+	public static function activate() {
+		global $wpdb;
+
+		$table_name = self::get_table_name();
+
+		// Create the table if it doesn't exist.
+		// @TODO: remove this custom SQLite declaration after first phase of unit tests is done.
+		if ( self::is_sqlite() ) {
+			$sql = $wpdb->prepare(
+				'CREATE TABLE IF NOT EXISTS %i (
+					id INTEGER PRIMARY KEY AUTOINCREMENT,
+					session_id INTEGER NOT NULL,
+					element_type INTEGER NOT NULL default %d,
+					element_id TEXT NOT NULL,
+					parent_id TEXT DEFAULT NULL,
+					byte_offset INTEGER NOT NULL,
+					sort_order int DEFAULT 1
+				);
+
+				CREATE UNIQUE INDEX IF NOT EXISTS idx_element_id ON %i (element_id);
+				CREATE INDEX IF NOT EXISTS idx_session_id ON %i (session_id);
+				CREATE INDEX IF NOT EXISTS idx_parent_id ON %i (parent_id);
+				CREATE INDEX IF NOT EXISTS idx_byte_offset ON %i (byte_offset);',
+				$table_name,
+				self::ELEMENT_TYPE_POST,
+				$table_name,
+				$table_name,
+				$table_name,
+				$table_name
+			);
+		} else {
+			// See wp_get_db_schema
+			$max_index_length = 191;
+
+			// MySQL, MariaDB.
+			$sql = $wpdb->prepare(
+				'CREATE TABLE IF NOT EXISTS %i (
+					id bigint(20) unsigned NOT NULL AUTO_INCREMENT,
+					session_id bigint(20) unsigned NOT NULL,
+					element_type tinyint(1) NOT NULL default %d,
+					element_id text NOT NULL,
+					parent_id text DEFAULT NULL,
+					byte_offset bigint(20) unsigned NOT NULL,
+					sort_order int DEFAULT 1,
+					PRIMARY KEY  (id),
+					KEY session_id (session_id),
+					KEY element_id (element_id(%d)),
+					KEY parent_id (parent_id(%d)),
+					KEY byte_offset (byte_offset)
+				) ' . $wpdb->get_charset_collate(),
+				self::get_table_name(),
+				self::ELEMENT_TYPE_POST,
+				$max_index_length,
+				$max_index_length
+			);
+		}
+
+		require_once ABSPATH . 'wp-admin/includes/upgrade.php';
+		dbDelta( $sql );
+
+		update_option( self::OPTION_NAME, self::DB_VERSION );
+	}
+
+	public static function is_sqlite() {
+		return defined( 'DB_ENGINE' ) && 'sqlite' === DB_ENGINE;
+	}
+
+	/**
+	 * Run in the 'plugins_loaded' action.
+	 */
+	public static function load() {
+		if ( self::DB_VERSION !== (int) get_site_option( self::OPTION_NAME ) ) {
+			// Used to update the database with dbDelta, if needed in the future.
+			self::activate();
+		}
+	}
+
+	/**
+	 * Run by register_deactivation_hook.
+	 */
+	public static function deactivate() {
```

Comment:
Would there be any advantage to running this when the plugin is uninstalled and not just deactivated?

Reply:
I've read this:

> If the users of your plugin do not need the data in this table if they deactivate the plugin, you could trigger this on the [plugin deactivation hook](https://developer.wordpress.org/reference/functions/register_deactivation_hook/).

register_deactivation_hook( __FILE__, 'wp_learn_delete_table' );
However, if the data in that table is important, and your users might want to keep it, even if the plugin is deactivated, you could delete the table using one of the [two uninstall methods](https://developer.wordpress.org/plugins/plugin-basics/uninstall-methods/) available to plugins.

Our data is not "important" in the strict sense of the term. It can be recreated again; it's just bulky. So, removing it when the user deactivates the plugin is okay. What do you think?

Reply:
Sounds good 👍